sysconfig/cloudinit.go

```diff
@@ -83,18 +83,112 @@ type supportedFilteredCloudConfig struct {
 
 type supportedFilteredDatasource struct {
 	// these are for MAAS
-	ConsumerKey string `yaml:"consumer_key"`
-	MetadataURL string `yaml:"metadata_url"`
-	TokenKey    string `yaml:"token_key"`
-	TokenSecret string `yaml:"token_secret"`
+	ConsumerKey string `yaml:"consumer_key,omitempty"`
+	MetadataURL string `yaml:"metadata_url,omitempty"`
+	TokenKey    string `yaml:"token_key,omitempty"`
+	TokenSecret string `yaml:"token_secret,omitempty"`
 }
 
 type supportedFilteredReporting struct {
-	Type        string `yaml:"type"`
-	Endpoint    string `yaml:"endpoint"`
-	ConsumerKey string `yaml:"consumer_key"`
-	TokenKey    string `yaml:"token_key"`
-	TokenSecret string `yaml:"token_secret"`
+	Type        string `yaml:"type,omitempty"`
+	Endpoint    string `yaml:"endpoint,omitempty"`
+	ConsumerKey string `yaml:"consumer_key,omitempty"`
+	TokenKey    string `yaml:"token_key,omitempty"`
+	TokenSecret string `yaml:"token_secret,omitempty"`
+}
+
+// filterCloudCfg filters a cloud-init configuration struct parsed from a single
+// cloud-init configuration file. The config provided here may be a subset of
+// the full cloud-init configuration from the file in that there may be
+// top-level keys in the YAML file that we did not parse and as such they are
+// dropped and filtered automatically. For other keys, we must parse part of the
+// configuration struct and remove nested keys while keeping other parts of the
+// same section.
+func filterCloudCfg(cfg *supportedFilteredCloudConfig, allowedDatasources []string) error {
```

Comment:
(nitpick) it looks like we never have an error condition here so we might as well not have an error return. But it's fine for now, if it's really not needed we could remove it in a followup.